internal/blobstore/swift.go

```diff
@@ -8,71 +8,129 @@ import (
+			AuthUrl:    params.AuthURL,
+			Region:     params.Region,
+			Tenant:     params.Tenant,
+			UserName:   params.Username,
+			StorageUrl: params.EndpointURL,
+		},
+		container: params.Bucket,
 	}
 }
 
 func (s *swiftBackend) Get(name string) (r ReadSeekCloser, size int64, err error) {
-	// Use infinite read-ahead here as the goose implementation of
-	// byte range handling seems to differ from swift's.
-	r2, headers, err := s.client.OpenObject(s.container, name, -1)
+	f, headers, err := s.connection.ObjectOpen(s.container, name, true, nil)
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if isNotFound(err) {
 			return nil, 0, errgo.WithCausef(nil, ErrNotFound, "")
 		}
 		return nil, 0, errgo.Mask(err)
 	}
-	lengthstr := headers.Get("Content-Length")
+	lengthstr := headers["Content-Length"]
 	size, err = strconv.ParseInt(lengthstr, 10, 64)
-	return swiftBackendReader{r2.(ReadSeekCloser)}, size, err
+	return swiftBackendReader{f}, size, errgo.Mask(err)
 }
 
+const maxBufferSize = 500 * 1024
+
 func (s *swiftBackend) Put(name string, r io.Reader, size int64, hash string) error {
-	h := NewHash()
-	r2 := io.TeeReader(r, h)
-	err := s.client.PutReader(s.container, name, r2, size)
-	if err != nil {
-		// TODO: investigate if PutReader can return err but the object still be
-		// written. Should there be cleanup here?
+	// Ensure that we never try to write too much data.
+	r = io.LimitReader(r, size)
+	// Buffer the first few KiB in memory so that we can retry if there is an authentication failure.
+	bufferSize := int64(maxBufferSize)
+	if size < bufferSize {
+		bufferSize = size
+	}
+	buf := make([]byte, size)
+	if _, err := io.ReadFull(r, buf); err != nil {
 		return errgo.Mask(err)
 	}
-	if hash != fmt.Sprintf("%x", h.Sum(nil)) {
-		err := s.client.DeleteObject(s.container, name)
+
+	var f *swift.ObjectCreateFile
+	h := NewHash()
+	var err error
+	for i := 0; i < 2; i++ {
+		f, err = s.connection.ObjectCreate(s.container, name, true, "", "", swift.Headers{"Content-Length": fmt.Sprintf("%d", size)})
+		if errgo.Cause(err) == swift.AuthorizationFailed {
+			continue
+		}
+		if err != nil {
+			break
+		}
+		w := io.MultiWriter(f, h)
+		_, err = w.Write(buf)
+		if errgo.Cause(err) == swift.AuthorizationFailed {
+			f.Close()
+			f = nil
+			h.Reset()
+			continue
+		}
 		if err != nil {
-			logger.Errorf("could not delete object from container after a hash mismatch was detected: %v", err)
+			break
 		}
-		return errgo.New("hash mismatch")
+		// If we've written first part successfully we expect the remainder to complete.
+		_, err = io.Copy(w, r)
+		break
 	}
-	return nil
+	if f != nil {
+		err1 := f.Close()
+		if err == nil {
+			err = err1
+		}
+	}
+	if err == nil && fmt.Sprintf("%x", h.Sum(nil)) != hash {
+		err1 := s.connection.ObjectDelete(s.container, name)
+		if err1 != nil {
+			logger.Errorf("could not delete object from container after a hash mismatch was detected: %v", err1)
+		}
+		err = errgo.New("hash mismatch")
```

Comment:
Any chance of a test for this case?

Reply:
not easily, this hasn't really changed from the previous code so if it wasn't tested before it won't be now.